public/app/core/components/Select/FolderPicker.tsx

```diff
@@ -195,25 +143,120 @@ export class FolderPicker extends PureComponent<Props, State> {
+  }, [folder, initialFolderId]);
+
+  // initial values for dropdown
+  useAsync(async () => {
+    if (skipInitialLoad) {
+      const folder = await getInitialValues({
+        getFolder: getFolderById,
+        folderId: initialFolderId,
+        folderName: initialTitle,
+      });
+      setFolder(folder);
+    }
+
+    await loadInitialValue();
+  }, [skipInitialLoad, initialFolderId, initialTitle]);
+
+  useEffect(() => {
+    if (folder && folder.id === VALUE_FOR_ADD) {
+      setIsCreatingNew(true);
+    }
+  }, [folder]);
+
+  const onFolderChange = useCallback(
+    (newFolder: SelectableValue<number>, actionMeta: ActionMeta) => {
+      const value = newFolder.value;
+      if (value === VALUE_FOR_ADD) {
+        setFolder({
+          id: VALUE_FOR_ADD,
+          title: newFolder?.title,
+        });
+      } else {
+        if (!newFolder) {
+          newFolder = { value: 0, label: rootName };
+        }
 
-    this.setState(
-      {
-        folder,
-      },
-      () => {
-        // if this is not the same as our initial value notify parent
-        if (folder && folder.value !== initialFolderId) {
-          this.props.onChange({ id: folder.value!, title: folder.label! });
+        if (actionMeta.action === 'clear' && onClear) {
+          onClear();
+          return;
         }
+
+        setFolder(newFolder);
+        onChange({ id: newFolder.value!, title: newFolder.label! });
       }
-    );
-  };
+    },
+    [onChange, onClear, rootName]
+  );
 
-  render() {
-    const { folder } = this.state;
-    const { enableCreateNew, inputId, onClear } = this.props;
-    const isClearable = typeof onClear === 'function';
+  const createNewFolder = useCallback(
+    async (folderName: string) => {
+      const newFolder = await createFolder({ title: folderName });
+      let folder: SelectableValue<number> = { value: -1, label: 'Not created' };
 
+      if (newFolder.id > -1) {
+        appEvents.emit(AppEvents.alertSuccess, ['Folder Created', 'OK']);
+        folder = { value: newFolder.id, label: newFolder.title };
+
+        setFolder(newFolder);
+        onFolderChange(folder, { action: 'create-option', option: folder });
+      } else {
+        appEvents.emit(AppEvents.alertError, ['Folder could not be created']);
+      }
+
+      return folder;
+    },
+    [onFolderChange]
+  );
+
+  const onKeyDown = useCallback(
+    (event: React.KeyboardEvent) => {
+      switch (event.key) {
+        case 'Enter': {
+          createNewFolder(folder?.title!);
+          setIsCreatingNew(false);
+          break;
+        }
+        case 'Escape': {
+          setFolder({ value: 0, label: rootName });
+          setIsCreatingNew(false);
+        }
+      }
+    },
+    [createNewFolder, folder, rootName]
+  );
+
+  const onNewFolderChange = (e: FormEvent<HTMLInputElement>) => {
+    const value = e.currentTarget.value;
+    setFolder({ id: -1, title: value });
+  };
+
+  if (isCreatingNew) {
```

Comment:
No need to check if `customAdd` is set to true?

Reply:
Not necessary, because `isCreatingnew` is going to be true only if we have `VALUE_FOR_ADD` in the options (that only happens if we have `customAdd` set to true
 ```
useEffect(() => {
    if (folder && folder.id === VALUE_FOR_ADD) {
      setIsCreatingNew(true);
    }
  }, [folder]); useEffect(() => {
    if (folder && folder.id === VALUE_FOR_ADD) {
      setIsCreatingNew(true);
    }
  }, [folder]);
```